File: unity/LaserTag/Assets/Scripts/Messages.cs
using System;
using UnityEngine;

[Serializable]
public struct HelloMessage
{
    public string id;
    public string app_type;
    public string message;

    public HelloMessage(string id, string app_type, string message)
    {
        this.id = id;
        this.app_type = app_type;
        this.message = message;
    }
}

[Serializable]
public struct PingMessage
{
    public float timestamp;

    public PingMessage(float timestamp)
    {
        this.timestamp = timestamp;
    }
}

[Serializable]
public struct AprilTagDetection
{
    public int tag_id;
    public float tag_size_cm;
    public Matrix4x4 pose_matrix;
}

[Serializable]
public struct AprilTagDetectionsMessage
{
    public AprilTagDetection[] detections;
}

[Serializable]
public struct HeightmapMessage
{
    public float[] heightmap;
    public float[] sample_count;
    public int width;
    public int height;
    public float cell_size;
}

[Serializable]
public struct Box
{
    public string label;
    public Matrix4x4 pose_matrix;
    public Vector3 extents;
}

[Serializable]
public struct BoxesMessage
{
    public Box[] boxes;
}

[Serializable]
public struct ModelBasedCalibrationControlMessage
{
    public bool startCalibration; // true to start a new calibration session, false ends current session
}

[Serializable]
public struct HMDControllerPoseMessage
{
    public Matrix4x4 pose_matrix;
}

[Serializable]
public struct ModelBasedRegistrationMessage
{
    public Matrix4x4 registration_matrix;
    public Vector3[] oakd_controller_positions;
}

[Serializable]
public struct HMDAvatarPoseMessage
{
    public string id;
    public Matrix4x4 head_pose;
    public Matrix4x4 left_hand_pose;
    public Matrix4x4 right_hand_pose;

    public HMDAvatarPoseMessage(string id, Matrix4x4 head_pose, Matrix4x4 left_hand_pose, Matrix4x4 right_hand_pose)
    {
        this.id = id;
        this.head_pose = head_pose;
        this.left_hand_pose = left_hand_pose;
        this.right_hand_pose = right_hand_pose;
    }
}

[Serializable]
public struct HMDLaserFiredMessage
{
    public string id;
    public Matrix4x4 pose;
    public bool fast_beam_mode;

    public HMDLaserFiredMessage(string id, Matrix4x4 pose, bool fast_beam_mode)
    {
        this.id = id;
        this.pose = pose;
        this.fast_beam_mode = fast_beam_mode;
    }
}

File: src/python/networking/messages.py
from dataclasses import dataclass
import numpy as np
from typing import List


@dataclass
class HelloMessage:
    id: str
    app_type: str
    message: str

@dataclass
class PingMessage:
    timestamp: float

@dataclass
class AprilTagDetection:
    tag_id: int
    tag_size_cm: float
    pose_matrix: np.ndarray

@dataclass
class AprilTagDetectionsMessage:
  detections: List[AprilTagDetection]

@dataclass
class HeightmapMessage:
    heightmap: List[float]
    sample_count: List[float]
    width: int
    height: int
    cell_size: float

@dataclass
class Box:
    label: str
    pose_matrix: np.ndarray
    extents: np.ndarray

@dataclass
class BoxesMessage:
    boxes: List[Box]

@dataclass
class ModelBasedCalibrationControlMessage:
    startCalibration: bool

@dataclass
class HMDControllerPoseMessage:
    pose_matrix: np.ndarray

@dataclass
class VuforiaControllerPoseMessage:
    pose_matrix: np.ndarray

@dataclass
class ModelBasedRegistrationMessage:
    registration_matrix: np.ndarray
    oakd_controller_positions: List[np.ndarray]

@dataclass
class HMDAvatarPoseMessage:
    id: str
    head_pose: np.ndarray
    left_hand_pose: np.ndarray
    right_hand_pose: np.ndarray

@dataclass
class HMDLaserFiredMessage:
    id: str
    pose: np.ndarray
    fast_beam_mode: bool